{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import rpy2.robjects as rob\n",
    "\n",
    "use_imputation = 1 # currently only 1\n",
    "\n",
    "eval_phenotype = 0\n",
    "# 0: PET av45\n",
    "# 1: CSF abeta\n",
    "# 2: CSF tau\n",
    "# 3: CSF ptau\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A TAB separated clmuped variant list. p<5e-5 variants, obtained from GWAS of iPSC samples.\n",
    "Base_clump = pd.read_csv(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/genotype/20190722_AD_HC_indvselect_qc_merge_imp_QC_excludedup_namemod.clumped_tab\",sep=\"\\t\")\n",
    "\n",
    "# Imputed genotype of iPSC samples\n",
    "Base_VCFfile = open(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/genotype/20190722_AD_HC_indvselect_qc_merge_imp_QC_excludedup_namemod.vcf\", mode=\"r\")\n",
    "\n",
    "# A file where the results will be written.\n",
    "Base_select_file = open(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/genotype/20190722_AD_HC_indvselect_qc_merge_imp_QC_selectedVCF_clump.txt\", mode=\"w\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished: line 100000\n",
      "finished: line 200000\n",
      "finished: line 300000\n",
      "finished: line 400000\n",
      "finished: line 500000\n",
      "finished: line 600000\n",
      "finished: line 700000\n",
      "finished: line 800000\n",
      "finished: line 900000\n",
      "finished: line 1000000\n",
      "finished: line 1100000\n",
      "finished: line 1200000\n",
      "finished: line 1300000\n",
      "finished: line 1400000\n",
      "finished: line 1500000\n",
      "finished: line 1600000\n",
      "finished: line 1700000\n",
      "finished: line 1800000\n",
      "finished: line 1900000\n",
      "finished: line 2000000\n",
      "finished: line 2100000\n",
      "finished: line 2200000\n",
      "finished: line 2300000\n",
      "finished: line 2400000\n",
      "finished: line 2500000\n",
      "finished: line 2600000\n",
      "finished: line 2700000\n",
      "finished: line 2800000\n",
      "finished: line 2900000\n",
      "finished: line 3000000\n",
      "finished: line 3100000\n",
      "finished: line 3200000\n",
      "finished: line 3300000\n",
      "finished: line 3400000\n",
      "finished: line 3500000\n",
      "finished: line 3600000\n",
      "finished: line 3700000\n",
      "finished: line 3800000\n",
      "finished: line 3900000\n",
      "finished: line 4000000\n",
      "finished: line 4100000\n",
      "finished: line 4200000\n",
      "finished: line 4300000\n",
      "finished: line 4400000\n",
      "finished: line 4500000\n",
      "finished: line 4600000\n",
      "finished: line 4700000\n",
      "finished: line 4800000\n",
      "finished: line 4900000\n",
      "finished: line 5000000\n",
      "finished: line 5100000\n",
      "finished: line 5200000\n",
      "finished: line 5300000\n",
      "finished: line 5400000\n",
      "finished: line 5500000\n",
      "finished: line 5600000\n",
      "finished: line 5700000\n",
      "finished: line 5800000\n",
      "finished: line 5900000\n",
      "finished: line 6000000\n",
      "finished: line 6100000\n",
      "finished: line 6200000\n",
      "finished: line 6300000\n",
      "finished: line 6400000\n",
      "finished: line 6500000\n",
      "finished: line 6600000\n",
      "finished: line 6700000\n",
      "finished: line 6800000\n",
      "finished: line 6900000\n",
      "finished: line 7000000\n",
      "finished: line 7100000\n",
      "finished: line 7200000\n",
      "finished: line 7300000\n"
     ]
    }
   ],
   "source": [
    "# Extract genotypes of selected variants from iPSC sample VCF and write extracted genotypes to a new file.\n",
    "\n",
    "## skip contig lines of the VCF file.\n",
    "cnt = 0\n",
    "for ind in range(0,29):\n",
    "    line = Base_VCFfile.readline()\n",
    "    cnt = cnt+1\n",
    "    \n",
    "line = Base_VCFfile.readline()\n",
    "GWAS_SNP_np = Base_clump[\"SNP\"].values\n",
    "while line:\n",
    "    feature = line.split(sep=\"\\t\")\n",
    "    if feature[2] in GWAS_SNP_np:\n",
    "        if len(np.where(GWAS_SNP_np==feature[2])) == 1:\n",
    "            Base_select_file.write(line)\n",
    "        else:\n",
    "            for ind in range(0,len(np.where(GWAS_SNP_np==feature[2]))):\n",
    "                if int(feature[0]) == int(GWAS_SNPlist[GWAS_SNP_np==feature[2]][0][ind]):\n",
    "                                          Base_select_file.write(line)\n",
    "\n",
    "    cnt = cnt+1\n",
    "    line = Base_VCFfile.readline()\n",
    "    if cnt%100000 == 0:\n",
    "        print(\"finished: line \"+str(cnt))\n",
    "\n",
    "Base_select_file.close()    \n",
    "Base_VCFfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished: line 100000\n",
      "finished: line 200000\n",
      "finished: line 300000\n",
      "finished: line 400000\n",
      "finished: line 500000\n",
      "finished: line 600000\n",
      "finished: line 700000\n",
      "finished: line 800000\n",
      "finished: line 900000\n",
      "finished: line 1000000\n",
      "finished: line 1100000\n",
      "finished: line 1200000\n",
      "finished: line 1300000\n",
      "finished: line 1400000\n",
      "finished: line 1500000\n",
      "finished: line 1600000\n",
      "finished: line 1700000\n",
      "finished: line 1800000\n",
      "finished: line 1900000\n",
      "finished: line 2000000\n",
      "finished: line 2100000\n",
      "finished: line 2200000\n",
      "finished: line 2300000\n",
      "finished: line 2400000\n",
      "finished: line 2500000\n",
      "finished: line 2600000\n",
      "finished: line 2700000\n",
      "finished: line 2800000\n",
      "finished: line 2900000\n",
      "finished: line 3000000\n",
      "finished: line 3100000\n",
      "finished: line 3200000\n",
      "finished: line 3300000\n",
      "finished: line 3400000\n",
      "finished: line 3500000\n",
      "finished: line 3600000\n",
      "finished: line 3700000\n",
      "finished: line 3800000\n",
      "finished: line 3900000\n",
      "finished: line 4000000\n",
      "finished: line 4100000\n",
      "finished: line 4200000\n",
      "finished: line 4300000\n",
      "finished: line 4400000\n",
      "finished: line 4500000\n",
      "finished: line 4600000\n",
      "finished: line 4700000\n",
      "finished: line 4800000\n",
      "finished: line 4900000\n",
      "finished: line 5000000\n",
      "finished: line 5100000\n",
      "finished: line 5200000\n",
      "finished: line 5300000\n",
      "finished: line 5400000\n",
      "finished: line 5500000\n",
      "finished: line 5600000\n",
      "finished: line 5700000\n",
      "finished: line 5800000\n",
      "finished: line 5900000\n",
      "finished: line 6000000\n",
      "finished: line 6100000\n",
      "finished: line 6200000\n",
      "finished: line 6300000\n",
      "finished: line 6400000\n",
      "finished: line 6500000\n",
      "finished: line 6600000\n",
      "finished: line 6700000\n",
      "finished: line 6800000\n",
      "finished: line 6900000\n",
      "finished: line 7000000\n",
      "finished: line 7100000\n",
      "finished: line 7200000\n",
      "finished: line 7300000\n",
      "finished: line 7400000\n",
      "finished: line 7500000\n",
      "finished: line 7600000\n",
      "finished: line 7700000\n",
      "finished: line 7800000\n",
      "finished: line 7900000\n",
      "finished: line 8000000\n",
      "finished: line 8100000\n",
      "finished: line 8200000\n",
      "finished: line 8300000\n",
      "finished: line 8400000\n",
      "finished: line 8500000\n",
      "finished: line 8600000\n",
      "finished: line 8700000\n",
      "finished: line 8800000\n",
      "finished: line 8900000\n",
      "finished: line 9000000\n",
      "finished: line 9100000\n",
      "finished: line 9200000\n",
      "finished: line 9300000\n",
      "finished: line 9400000\n",
      "finished: line 9500000\n",
      "finished: line 9600000\n",
      "finished: line 9700000\n",
      "finished: line 9800000\n",
      "finished: line 9900000\n",
      "finished: line 10000000\n",
      "finished: line 10100000\n"
     ]
    }
   ],
   "source": [
    "# Extract genotypes of selected variants from ADNI sample VCF and write extracted genotypes to a new file.\n",
    "\n",
    "# imputed ADNI VCF file. \n",
    "Target_VCFfile = open(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/ADNI/WGS_Omni25_merge_imp_QC_excludedup_namemod.vcf\", mode=\"r\")\n",
    "# A file where the results will be written.\n",
    "Target_select_file = open(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/ADNI/WGS_Omni25_merge_imp_QC_selectedVCF_clump.txt\", mode=\"w\") \n",
    "    \n",
    "\n",
    "## skip contig lines of the VCF file.\n",
    "cnt = 0\n",
    "for ind in range(0,29):\n",
    "    line = Target_VCFfile.readline()\n",
    "    cnt = cnt+1\n",
    "    \n",
    "line = Target_VCFfile.readline()\n",
    "\n",
    "while line:\n",
    "    feature = line.split(sep=\"\\t\")\n",
    "    if feature[2] in GWAS_SNP_np:\n",
    "        if len(np.where(GWAS_SNP_np==feature[2])) == 1:\n",
    "            Target_select_file.write(line)\n",
    "        else:\n",
    "            for ind in range(0,len(np.where(GWAS_SNP_np==feature[2]))):\n",
    "                if int(feature[0]) == int(GWAS_SNPlist[GWAS_SNP_np==feature[2]][0][ind]):\n",
    "                                          Target_select_file.write(line)\n",
    "    cnt = cnt+1\n",
    "    line = Target_VCFfile.readline()\n",
    "    if cnt%100000 == 0:\n",
    "        print(\"finished: line \"+str(cnt))\n",
    "\n",
    "Target_select_file.close()    \n",
    "Target_VCFfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Re-load previously wrote files.\n",
    "Base_selectedSNPlist = pd.read_csv(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/genotype/20190722_AD_HC_indvselect_qc_merge_imp_QC_selectedVCF_clump.txt\",sep=\"\\t\",header=None)\n",
    "Target_selectedSNPlist = pd.read_csv(\"/Users/yuichiro/Workspace/Genome/ADabetaQTL/ADNI/WGS_Omni25_merge_imp_QC_selectedVCF_clump.txt\",sep=\"\\t\",header=None)\n",
    "\n",
    "base_selected_var = Base_selectedSNPlist[2].tolist()\n",
    "target_selected_var = Target_selectedSNPlist[2].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Base_selectedSNPlist' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-628328580767>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     11\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mgeno_num\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m \u001b[0mbase_geno_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mBase_selectedSNPlist\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m9\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;36m110\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapplymap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenotype_format2num\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# 102 samples in iPSC data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0mbase_geno_np\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbase_geno_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Base_selectedSNPlist' is not defined"
     ]
    }
   ],
   "source": [
    "# loading genotypes to numpy array.\n",
    "import subprocess\n",
    "\n",
    "def genotype_format2num(geno_with_pipe_column):\n",
    "    if use_imputation == 1:\n",
    "        geno_with_pipe = geno_with_pipe_column.split(\"/\")\n",
    "        if geno_with_pipe[0].isdecimal() and geno_with_pipe[1].isdecimal():\n",
    "            geno_num = int(geno_with_pipe[0]) + int(geno_with_pipe[1])\n",
    "        else:\n",
    "            geno_num = np.nan\n",
    "    return geno_num\n",
    "    \n",
    "# load iPSC sample gentypes to numpy array\n",
    "base_geno_df = Base_selectedSNPlist.loc[:,9:110].applymap(genotype_format2num) # 102 samples in iPSC data\n",
    "base_geno_np = base_geno_df.values.transpose()\n",
    "\n",
    "# load ADNI sample gentypes to numpy array\n",
    "target_geno_df = Target_selectedSNPlist.loc[:,9:723].applymap(genotype_format2num) # 715 samples in ADNI data\n",
    "if use_imputation == 1:\n",
    "    Target_VCFfile = open(\"ADNI/WGS_Omni25_merge_imp_QC_excludedup_namemod.vcf\", mode=\"r\")\n",
    "    \n",
    "## skip contig lines of the VCF file.\n",
    "cnt = 0\n",
    "for ind in range(0,28):\n",
    "    line = Target_VCFfile.readline()\n",
    "    cnt = cnt+1\n",
    "            \n",
    "line = Target_VCFfile.readline()\n",
    "feature = line.split(sep=\"\\t\")\n",
    "target_VCFsample_list = feature[9:724]\n",
    "target_VCFsample_list[714] = target_VCFsample_list[714].split(sep=\"\\n\")[0]\n",
    "Target_VCFfile.close()\n",
    "if eval_phenotype==0:\n",
    "    target_pheno = pd.read_csv(\"pheno_merge_ADNI_SUVR.txt\",sep=\"\\t\")\n",
    "else:\n",
    "    target_pheno = pd.read_csv(\"pheno_merge_ADNI_CSF.txt\",sep=\"\\t\")\n",
    "target_phenosample_list = []\n",
    "cnt=0\n",
    "for index, row in target_pheno.iterrows():\n",
    "    target_phenosample_list.append(str(int(row[\"FID\"]))+\"_\"+str(int(row[\"IID\"])).zfill(4))\n",
    "    cnt=cnt+1\n",
    "\n",
    "\n",
    "target_df_keep = []\n",
    "cnt=0\n",
    "for samp in target_VCFsample_list:\n",
    "    if samp in target_phenosample_list:\n",
    "        target_df_keep.append(cnt)\n",
    "    cnt = cnt+1\n",
    "\n",
    "target_geno_df = target_geno_df.filter(target_geno_df.columns[target_df_keep], axis=1)\n",
    "target_geno_np = target_geno_df.values.transpose()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(581, 3)\n",
      "(581,)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAASXUlEQVR4nO3deZBlZX3G8e8DiBsahtBSEyA2IomSqKOOI4oacAWtCKYIES1FQxyrXCLBmCJqJZrkD61ySUwiZhQUUwbcoFwwGETKXUiD7LggAQUR2gVFTKIDv/xxz8RLTy93pufc2z3v91N1q899zzn3/N6Bevrct895T6oKSVI7dpl0AZKk8TL4JakxBr8kNcbgl6TGGPyS1JjdJl3AKPbee++anp6edBmStKpcfPHFP6iqqbntqyL4p6enmZmZmXQZkrSqJLlhvnaHeiSpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTGr4s5dbZvpk8+ZyHGvf9OzJnJcSdvGM35JaozBL0mN6S34k9wryUVJLktyVZI3du0HJLkwybVJPphk975qkCRtrc8z/v8FnlxVjwDWAUckOQR4M/D2qnow8GPghB5rkCTN0Vvw18DPurf36F4FPBn4SNd+OnB0XzVIkrbW6xh/kl2TXArcCpwHfBu4rao2d5vcCOy7wL4bk8wkmZmdne2zTElqSq/BX1V3VtU6YD9gA/CQbdh3U1Wtr6r1U1NbPUBGkrSdxnJVT1XdBlwAPA7YM8mW+wf2A24aRw2SpIE+r+qZSrJnt3xv4GnANQx+ARzTbXY88LG+apAkba3PO3fXAqcn2ZXBL5gPVdUnk1wNnJnk74CvAaf2WIMkaY7egr+qLgceOU/7dQzG+yVJE+Cdu5LUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TG+Mxd7TCTetYv+LxfaVt4xi9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JakxvwZ9k/yQXJLk6yVVJXtW1vyHJTUku7V7P7KsGSdLW+nwC12bg1VV1SZL7ARcnOa9b9/aqekuPx5YkLaC34K+qm4Gbu+Xbk1wD7NvX8SRJoxnLGH+SaeCRwIVd0yuSXJ7ktCRrFthnY5KZJDOzs7PjKFOSmtB78CfZA/gocGJV/RQ4BTgQWMfgG8Fb59uvqjZV1fqqWj81NdV3mZLUjF6DP8k9GIT+B6rqLICquqWq7qyqu4B3Axv6rEGSdHd9XtUT4FTgmqp621D72qHNngNc2VcNkqSt9XlVz6HAC4Arklzatb0WOC7JOqCA64GX9liDJGmOPq/q+SKQeVZ9qq9jSpKW5p27ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4JekxvQW/En2T3JBkquTXJXkVV37XknOS/Kt7ueavmqQJG2tzzP+zcCrq+pg4BDg5UkOBk4Gzq+qg4Dzu/eSpDHpLfir6uaquqRbvh24BtgXOAo4vdvsdODovmqQJG1tLGP8SaaBRwIXAvtU1c3dqu8D+yywz8YkM0lmZmdnx1GmJDWh9+BPsgfwUeDEqvrp8LqqKqDm26+qNlXV+qpaPzU11XeZktSMXoM/yT0YhP4HquqsrvmWJGu79WuBW/usQZJ0d31e1RPgVOCaqnrb0KqPA8d3y8cDH+urBknS1nbr8bMPBV4AXJHk0q7ttcCbgA8lOQG4ATi2xxokSXP0FvxV9UUgC6x+Sl/HlSQtzjt3JakxBr8kNcbgl6TGGPyS1BiDX5IaY/BLUmMMfklqzEjBn+TQUdokSSvfqGf8/zhimyRphVv0zt0kjwMeD0wlOWlo1f2BXfssTJLUj6WmbNgd2KPb7n5D7T8FjumrKElSfxYN/qr6HPC5JO+rqhvGVJMkqUejTtJ2zySbgOnhfarqyX0UJUnqz6jB/2HgXcB7gDv7K0eS1LdRg39zVZ3SayU7oemTz5l0CZK0lVEv5/xEkpclWZtkry2vXiuTJPVi1DP+LY9KfM1QWwEP2rHlSJL6NlLwV9UBfRciSRqPkYI/yQvna6+q9+/YciRJfRt1qOcxQ8v3YvDM3EsAg1+SVplRh3peOfw+yZ7AmX0UJEnq1/ZOy3wH4Li/JK1Co47xf4LBVTwwmJztocCH+ipKktSfUcf43zK0vBm4oapu7KEeSVLPRhrq6SZr+zqDGTrXAL9Yap8kpyW5NcmVQ21vSHJTkku71zO3t3BJ0vYZ9QlcxwIXAX8IHAtcmGSpaZnfBxwxT/vbq2pd9/rUthQrSVq+UYd6Xgc8pqpuBUgyBXwG+MhCO1TV55NML7tCSdIONepVPbtsCf3OD7dh37lekeTybihozXZ+hiRpO40a3ucm+XSSFyV5EXAOsD3DNKcABwLrgJuBty60YZKNSWaSzMzOzm7HoSRJ81k0+JM8OMmhVfUa4F+Ah3evrwCbtvVgVXVLVd1ZVXcB7wY2LLLtpqpaX1Xrp6amtvVQkqQFLHXG//cMnq9LVZ1VVSdV1UnA2d26bZJk7dDb5wBXLrStJKkfS/1xd5+qumJuY1VdsdQfbpOcARwG7J3kRuCvgcOSrGNwM9j1wEu3vWRJ0nIsFfx7LrLu3ovtWFXHzdN86lIFSZL6tdRQz0ySl8xtTPInwMX9lCRJ6tNSZ/wnAmcneT6/Cvr1wO4MxuglSavMosFfVbcAj09yOPC7XfM5VfXZ3iuTJPVi1Pn4LwAu6LkWSdIYbO/dt5KkVcrgl6TGGPyS1BiDX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMYY/JLUGINfkhpj8EtSY3oL/iSnJbk1yZVDbXslOS/Jt7qfa/o6viRpfn2e8b8POGJO28nA+VV1EHB+916SNEa9BX9VfR740Zzmo4DTu+XTgaP7Or4kaX7jHuPfp6pu7pa/D+yz0IZJNiaZSTIzOzs7nuokqQET++NuVRVQi6zfVFXrq2r91NTUGCuTpJ3buIP/liRrAbqft475+JLUvHEH/8eB47vl44GPjfn4ktS8Pi/nPAP4CvDbSW5McgLwJuBpSb4FPLV7L0kao936+uCqOm6BVU/p65iSpKV5564kNaa3M35pnKZPPmcix73+Tc+ayHGl5fCMX5IaY/BLUmMMfklqjMEvSY0x+CWpMQa/JDXG4Jekxhj8ktQYg1+SGmPwS1JjDH5JaozBL0mNcZI2aRkmNTkcOEGctp9n/JLUGINfkhpj8EtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGGPyS1JiJ3Lmb5HrgduBOYHNVrZ9EHZLUoklO2XB4Vf1ggseXpCY51CNJjZlU8BfwH0kuTrJxvg2SbEwyk2RmdnZ2zOVJ0s5rUsH/hKp6FHAk8PIkT5q7QVVtqqr1VbV+ampq/BVK0k5qIsFfVTd1P28FzgY2TKIOSWrR2IM/yX2T3G/LMvB04Mpx1yFJrZrEVT37AGcn2XL8f6uqcydQhyQ1aezBX1XXAY8Y93ElSQNezilJjfGZu5K0hJ3t2cqe8UtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TG7PSXc07yMixJWok845ekxhj8ktQYg1+SGmPwS1JjDH5JasxOf1WPpB1rZ5uwrEWe8UtSYwx+SWqMwS9JjTH4JakxBr8kNcbgl6TGeDmnpFXDSRd3DM/4JakxBr8kNcbgl6TGTCT4kxyR5BtJrk1y8iRqkKRWjT34k+wK/DNwJHAwcFySg8ddhyS1ahJn/BuAa6vquqr6BXAmcNQE6pCkJk3ics59ge8Ovb8ReOzcjZJsBDZ2b3+W5BtjqA1gb+AHYzrWStJqv2GV9j1vXvZHrMp+7wCrqt/L/O/8wPkaV+x1/FW1Cdg07uMmmamq9eM+7qS12m9ot+/2u12TGOq5Cdh/6P1+XZskaQwmEfz/CRyU5IAkuwPPBT4+gTokqUljH+qpqs1JXgF8GtgVOK2qrhp3HYsY+/DSCtFqv6HdvtvvRqWqJl2DJGmMvHNXkhpj8EtSY5oK/qWmikhyzyQf7NZfmGS6a39+kkuHXnclWTfu+rfXMvp9jySnJ7kiyTVJ/nLsxS/DMvq9e5L3dv2+LMlhYy59WUbo95OSXJJkc5Jj5qw7Psm3utfx46t6x1hm389NcluST46v4gmpqiZeDP6Q/G3gQcDuwGXAwXO2eRnwrm75ucAH5/mchwHfnnR/xtFv4HnAmd3yfYDrgelJ92kM/X458N5u+QHAxcAuk+7TDuz3NPBw4P3AMUPtewHXdT/XdMtrJt2ncfS9W/cU4PeBT066L32/WjrjH2WqiKOA07vljwBPSZI52xzX7btaLKffBdw3yW7AvYFfAD8dT9nLtpx+Hwx8FqCqbgVuA1bLDT9L9ruqrq+qy4G75uz7DOC8qvpRVf0YOA84YhxF7yDL6TtVdT5w+1gqnbCWgn++qSL2XWibqtoM/AT49Tnb/BFwRk819mE5/f4IcAdwM/Ad4C1V9aO+C95BltPvy4BnJ9ktyQHAo7n7TYcr2Sj97mPflWC11z82K3bKhpUoyWOBn1fVlZOuZUw2AHcCv8Hgq/8Xknymqq6bbFm9Ow14KDAD3AB8mcG/g7RTaOmMf5SpIv5/m25449eAHw6tfy6r62wfltfv5wHnVtUvuyGPL7F6hjy2u99Vtbmq/qyq1lXVUcCewDf7L3mHWM6UKKt9OpXVXv/YtBT8o0wV8XFgy5UMxwCfre6vPkl2AY5ldY3vw/L6/R3gyQBJ7gscAnx9LFUv33b3O8l9uv6S5GnA5qq6elyFL9NypkT5NPD0JGuSrAGe3rWtFk4HM6pJ/3V5nC/gmQzO3L4NvK5r+xvg2d3yvYAPA9cCFwEPGtr3MOCrk+7DOPsN7NG1XwVcDbxm0n0ZU7+ngW8A1wCfAR446b7s4H4/hsH49x0MvtldNbTvH3f/HtcCL550X8bc9y8As8B/d9s8Y9L96evllA2S1JiWhnokSRj8ktQcg1+SGmPwS1JjDH5JaozBrxUrydFJKslDhtoO29bZE5OcmOQ+21nDq7sa9u7ePz/J5d3MnV9O8og5278ryaGL7J8k7+hmj7w8yaO2p66F+pbkU0n2XM5naudn8GslOw74YvdzOU5kMLvoNkmyP4ObmL4z1PxfwO9V1cOAv2Xrx/gdAnx1kf2PBA7qXhuBU5aoId3Ngws5kaG+VdUzq+q2xT5TMvi1IiXZA3gCcAKDOzCH3T/JOd286+/aEoxJnp7kK9186x9OskeSP2Uw19AFSS7otjslyUySq5K8cZEy3g78BYNZSgGoqi/XYOZKGAT8fkM1PxT4ZlXdudD+DGaLfH8NfBXYM8naOX2f7vr2fuBKYP/5al6gb9cPfbs4KcmV3evERfqpxhj8WqmOYjBP0DeBHyZ59NC6DcArGUyffCDwB13YvR54alU9isEEaydV1TuA7wGHV9Xh3f6vq6r1DOZl/70kD5978CRHATdV1WWL1HgC8O9D748Ezl1i/1FnkDwIeGdV/U5V3TBfzQv0bUv9jwZeDDyWwbeQlyR55CJ9UUOcnVMr1XHAP3TLZ3bvL+7eX1TdDKFJzmDwzeB/GPwi+FL3CIXdga8s8NnHJtnI4P//td1+l29Z2Y2Zv5bBMM28khzOIPifMNT8DODFo+w/ghu6bwQj1TyPJwBnV9UdXb1nAU8EvraMmrSTMPi14iTZi8HkcA9LUgyerFRJXtNtMneekQLC4CEii/49oJtf/8+Bx1TVj5O8j8GcPcMOBA4ALut+iewHXJJkQ1V9v/uG8B7gyKr6Yfe59wH2rKrvJXnYQvsz+gySd2xjzdLIHOrRSnQM8K9V9cCqmq6q/Rn8UfWJ3foN3QyMuzB4MM4XGYy3H5rkwTCYTTTJb3Xb3w7cr1u+P4NQ/UmSfRgMz9xNVV1RVQ/ojj3NYDjmUV3o/yZwFvCCbhhqi8OBC5ban8FskS/s/mh7CPCTqrp5iX+PxWoe7tuwLwBH51czjT6na5M849eKdBzw5jltH+3aP8hg+t1/Ah7MIGzPrqq7krwIOCPJPbt9Xs9gpsZNwLlJvldVhyf5GoPppb/L4BkD2+KvGDyl653d2fzmbuz9SAZPLFvKpxjMIHkt8HMG4/CLqqrLFqn5bn0b2ueS7pvBRV3Te6rKYR4BODuntCMkuQR4bFX9ctK1SEsx+CWpMY7xS1JjDH5JaozBL0mNMfglqTEGvyQ1xuCXpMb8H919fxg3MO9bAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEGCAYAAACKB4k+AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAU2klEQVR4nO3df5RfdX3n8edLUlDUmmDGNEviTlRql3VbZUcKpe0quIpgge5xMayrkeLm7Oq6/jq6IOfU7p7tOdB6itqzRVOhxpaClGKhYEWKoGf3aNwB5bdIRJBkgQxadKu71eh7/7h3br4Ok2Qy4fu9XzLPxzlz5t7Pvff7fc8nk7xyP/fezzdVhSRJAE/puwBJ0vgwFCRJHUNBktQxFCRJHUNBktRZ1ncB+2PlypU1OTnZdxmS9KRy8803P1pVE/Nte1KHwuTkJNPT032XIUlPKkke2N02h48kSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSZ0n9RPNktSnybOv7e297z/v5KG8rmcKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6gwtFJJcnGRHkjvm2fbuJJVkZbueJB9OsjXJbUmOGlZdkqTdG+aZwseBE+c2JlkLvBL41kDzq4Ej2q+NwIVDrEuStBtDC4Wq+gLwnXk2XQC8F6iBtlOBT1TjS8DyJKuHVZskaX4jvaaQ5FRge1XdOmfT4cCDA+vb2rb5XmNjkukk0zMzM0OqVJKWppGFQpJDgfcBv70/r1NVm6pqqqqmJiYmnpjiJEnAaD957fnAOuDWJABrgFuSHA1sB9YO7LumbZMkjdDIzhSq6vaqek5VTVbVJM0Q0VFV9TBwNfDG9i6kY4DvVtVDo6pNktQY5i2plwJfBF6YZFuSs/aw+6eB+4CtwB8DbxlWXZKk3Rva8FFVnbGX7ZMDywW8dVi1SJIWxieaJUkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1DEUJEkdQ0GS1BnmZzRfnGRHkjsG2n4/ydeS3JbkU0mWD2w7J8nWJPckedWw6pIk7d4wzxQ+Dpw4p+164EVV9YvA14FzAJIcCawH/ml7zB8lOWiItUmS5jG0UKiqLwDfmdP22ara2a5+CVjTLp8KXFZV/1BV3wS2AkcPqzZJ0vz6vKbwW8DftMuHAw8ObNvWtj1Oko1JppNMz8zMDLlESVpaegmFJOcCO4FL9vXYqtpUVVNVNTUxMfHEFydJS9iyUb9hkjcBrwFOqKpqm7cDawd2W9O2SZJGaKRnCklOBN4LnFJVPxjYdDWwPskhSdYBRwBfHmVtkqQhnikkuRR4GbAyyTbg/TR3Gx0CXJ8E4EtV9e+r6s4klwN30QwrvbWqfjys2iRJ8xtaKFTVGfM0X7SH/X8X+N1h1SNJ2jufaJYkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVJnaKGQ5OIkO5LcMdB2WJLrk9zbfl/RtifJh5NsTXJbkqOGVZckafeGeabwceDEOW1nAzdU1RHADe06wKuBI9qvjcCFQ6xLkrQbQwuFqvoC8J05zacCm9vlzcBpA+2fqMaXgOVJVg+rNknS/EZ9TWFVVT3ULj8MrGqXDwceHNhvW9smSRqh3i40V1UBta/HJdmYZDrJ9MzMzBAqk6Sla9Sh8MjssFD7fUfbvh1YO7DfmrbtcapqU1VNVdXUxMTEUIuVpKVm1KFwNbChXd4AXDXQ/sb2LqRjgO8ODDNJkkZk2bBeOMmlwMuAlUm2Ae8HzgMuT3IW8ABwerv7p4GTgK3AD4Azh1WXJGn3hhYKVXXGbjadMM++Bbx1WLVIkhbGJ5olSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUMRQkSR1DQZLUWdA0F0mOq6r/ubc2aXcmz762l/e9/7yTe3lf6clqoWcKf7jANknSk9gezxSSHAv8CjCR5F0Dm34WOGiYhUmSRm9vw0cHA89o93vmQPv3gNcOqyhJUj/2GApV9Xng80k+XlUPjKgmSVJPFvp5Cock2QRMDh5TVccPoyhJUj8WGgp/AXwE+Bjw4+GVI0nq00JDYWdVXTjUSiRJvVvoLal/neQtSVYnOWz2a6iVSZJGbqFnChva7+8ZaCvgeYt50yTvBN7cvsbtwJnAauAy4NnAzcAbquqHi3l9SdLiLOhMoarWzfO12EA4HPhPwFRVvYjmeYf1wPnABVX1AuDvgLMW8/qSpMVb6DQXb5yvvao+sR/v+7QkPwIOBR4Cjgf+Tbt9M/A7gNcxJGmEFjp89NKB5acCJwC3APscClW1PckHgG8B/xf4LM1w0WNVtbPdbRtw+HzHJ9kIbAR47nOfu69vL0nagwWFQlW9bXA9yXKa8f99lmQFcCqwDniM5nbXExd6fFVtAjYBTE1N1WJqkCTNb7FTZ3+f5h/1xXgF8M2qmqmqHwFXAscBy5PMhtQaYPsiX1+StEgLvabw1zR3CkFzYfifAJcv8j2/BRyT5FCa4aMTgGngRpr5lC6judvpqkW+viRpkRZ6TeEDA8s7gQeqatti3rCqtiS5guaaxE7gKzTDQdcClyX5b23bRYt5fUnS4i30msLnk6xi1wXne/fnTavq/cD75zTfBxy9P68rSdo/C7qmkOR04MvAvwZOB7YkcepsSTrALHT46FzgpVW1AyDJBPC3wBXDKkySNHoLvfvoKbOB0Pr2PhwrSXqSWOiZwmeSXAdc2q6/Dvj0cEqSJPVlb5/R/AJgVVW9J8m/An613fRF4JJhFydJGq29nSl8EDgHoKqupHnQjCT/rN32G0OsTZI0Ynu7LrCqqm6f29i2TQ6lIklSb/YWCsv3sO1pT2AdkqQxsLdQmE7y7+Y2JnkzzcymkqQDyN6uKbwD+FSS17MrBKaAg4HfHGJdkqQe7DEUquoR4FeSvBx4Udt8bVV9buiVSZJGbqFzH91IM4upJOkA5lPJkqSOoSBJ6hgKkqSOoSBJ6ix0QjwdACbPvrbvEiSNOUNBB7Q+g/D+807u7b2lxepl+CjJ8iRXJPlakruTHJvksCTXJ7m3/b6ij9okaSnr65rCh4DPVNUvAL8E3A2cDdxQVUcAN7TrkqQRGnkoJHkW8OvARQBV9cOqegw4Fdjc7rYZOG3UtUnSUtfHmcI6YAb4kyRfSfKxJE+nmab7oXafh4FV8x2cZGOS6STTMzMzIypZkpaGPkJhGXAUcGFVvQT4PnOGiqqqgJrv4KraVFVTVTU1MTEx9GIlaSnpIxS2Aduqaku7fgVNSDySZDVA+31HD7VJ0pI28lCoqoeBB5O8sG06AbgLuBrY0LZtAK4adW2StNT19ZzC24BLkhwM3AecSRNQlyc5C3gAOL2n2iRpyeolFKrqqzQf1jPXCSMuRZI0wLmPJEkdQ0GS1HHuI2lI+pp3yTmXtD88U5AkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdXyiuQd9PekqSXvjmYIkqWMoSJI6hoIkqWMoSJI6hoIkqdNbKCQ5KMlXklzTrq9LsiXJ1iSfbD+/WZI0Qn2eKbwduHtg/Xzggqp6AfB3wFm9VCVJS1gvoZBkDXAy8LF2PcDxwBXtLpuB0/qoTZKWsr7OFD4IvBf4Sbv+bOCxqtrZrm8DDp/vwCQbk0wnmZ6ZmRl6oZK0lIw8FJK8BthRVTcv5viq2lRVU1U1NTEx8QRXJ0lLWx/TXBwHnJLkJOCpwM8CHwKWJ1nWni2sAbb3UJukRepz+pb7zzu5t/c+0Iz8TKGqzqmqNVU1CawHPldVrwduBF7b7rYBuGrUtUnSUjdOzyn8Z+BdSbbSXGO4qOd6JGnJ6XWW1Kq6CbipXb4POLrPeiRpqXPqbOkA49Ts2h/jNHwkSeqZoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6hgKkqSOoSBJ6vghO5Ke9PxgoSfOyM8UkqxNcmOSu5LcmeTtbfthSa5Pcm/7fcWoa5Okpa6P4aOdwLur6kjgGOCtSY4EzgZuqKojgBvadUnSCI08FKrqoaq6pV3+P8DdwOHAqcDmdrfNwGmjrk2SlrpeLzQnmQReAmwBVlXVQ+2mh4FVfdUlSUtVb6GQ5BnAXwLvqKrvDW6rqgJqN8dtTDKdZHpmZmYElUrS0tFLKCT5GZpAuKSqrmybH0myut2+Gtgx37FVtamqpqpqamJiYjQFS9IS0cfdRwEuAu6uqj8Y2HQ1sKFd3gBcNeraJGmp6+M5heOANwC3J/lq2/Y+4Dzg8iRnAQ8Ap/dQmyQtaSMPhar6H0B2s/mEUdYiSfppTnMhSeoYCpKkjqEgSeos2QnxnEBLkh7PMwVJUsdQkCR1DAVJUsdQkCR1DAVJUsdQkCR1DAVJUsdQkCR1DAVJUsdQkCR1DAVJUsdQkCR1DAVJUsdQkCR1DAVJUmfsQiHJiUnuSbI1ydl91yNJS8lYhUKSg4D/DrwaOBI4I8mR/VYlSUvHWIUCcDSwtaruq6ofApcBp/ZckyQtGeP2cZyHAw8OrG8DfnlwhyQbgY3t6t8nuWeR77USeHSRx46KNe6/ca8Pxr/Gca8Pxr/GJ7y+nL9fh//j3W0Yt1DYq6raBGza39dJMl1VU09ASUNjjftv3OuD8a9x3OuD8a9x3OsbNG7DR9uBtQPra9o2SdIIjFso/C/giCTrkhwMrAeu7rkmSVoyxmr4qKp2JvmPwHXAQcDFVXXnkN5uv4egRsAa99+41wfjX+O41wfjX+O419dJVfVdgyRpTIzb8JEkqUeGgiSpc8CGQpK1SW5McleSO5O8vW0/LMn1Se5tv69o25Pkw+30GrclOWpEdR6U5CtJrmnX1yXZ0tbxyfaCO0kOade3ttsnR1Tf8iRXJPlakruTHDtOfZjkne2f7x1JLk3y1L77MMnFSXYkuWOgbZ/7LMmGdv97k2wYQY2/3/4535bkU0mWD2w7p63xniSvGmgfyrQ089U3sO3dSSrJynZ9bPqwbX9b2493Jvm9gfaR9uGiVdUB+QWsBo5ql58JfJ1m6ozfA85u288Gzm+XTwL+BghwDLBlRHW+C/hz4Jp2/XJgfbv8EeA/tMtvAT7SLq8HPjmi+jYDb26XDwaWj0sf0jzs+E3gaQN996a++xD4deAo4I6Btn3qM+Aw4L72+4p2ecWQa3wlsKxdPn+gxiOBW4FDgHXAN2huBDmoXX5e+7txK3DksOpr29fS3IjyALByDPvw5cDfAoe068/pqw8X/XP1+eYj/UHhKuBfAvcAq9u21cA97fJHgTMG9u/2G2JNa4AbgOOBa9pf6kcH/mIeC1zXLl8HHNsuL2v3y5DrexbNP7qZ0z4WfciuJ+APa/vkGuBV49CHwOScfyz2qc+AM4CPDrT/1H7DqHHOtt8ELmmXzwHOGdh2XduvXd/Ot98w6gOuAH4JuJ9doTA2fUjzH5JXzLNfL324mK8DdvhoUDtM8BJgC7Cqqh5qNz0MrGqX55ti4/Ahl/ZB4L3AT9r1ZwOPVdXOeWro6mu3f7fdf5jWATPAn7RDXB9L8nTGpA+rajvwAeBbwEM0fXIz49WHs/a1z/r4fRz0WzT/+2YPtYy0xiSnAtur6tY5m8aivtbPA7/WDk9+PslLx7DGPTrgQyHJM4C/BN5RVd8b3FZNNPdyT26S1wA7qurmPt5/gZbRnB5fWFUvAb5PM/TR6bkPV9BMmLgO+EfA04ET+6hlX/TZZwuR5FxgJ3BJ37XMSnIo8D7gt/uuZS+W0Zy5HgO8B7g8Sfotad8c0KGQ5GdoAuGSqrqybX4kyep2+2pgR9s+6ik2jgNOSXI/zWywxwMfApYnmX2ocLCGrr52+7OAbw+xPmj+17Ktqra061fQhMS49OErgG9W1UxV/Qi4kqZfx6kPZ+1rn/Uy5UuSNwGvAV7fhte41Ph8mvC/tf07swa4JcnPjUl9s7YBV1bjyzSjACvHrMY9OmBDoU3ni4C7q+oPBjZdDczehbCB5lrDbPsb2zsZjgG+O3C6/4SrqnOqak1VTdJc9PxcVb0euBF47W7qm637te3+Q/3fZlU9DDyY5IVt0wnAXYxJH9IMGx2T5ND2z3u2vrHpwwH72mfXAa9MsqI9I3pl2zY0SU6kGc48pap+MKf29Wnu3loHHAF8mRFOS1NVt1fVc6pqsv07s43mRpKHGaM+BP6K5mIzSX6e5uLxo4xBHy5Ynxc0hvkF/CrNKfptwFfbr5NoxpBvAO6luUvgsHb/0HzAzzeA24GpEdb6MnbdffQ8ml+WrcBfsOsuhqe261vb7c8bUW0vBqbbfvwrmrs4xqYPgf8CfA24A/hTmrs7eu1D4FKaaxw/ovnH66zF9BnNuP7W9uvMEdS4lWZ8e/bvy0cG9j+3rfEe4NUD7SfR3Nn3DeDcYdY3Z/v97LrQPE59eDDwZ+3v4y3A8X314WK/nOZCktQ5YIePJEn7zlCQJHUMBUlSx1CQJHUMBUlSx1CQFijJue3Ml7cl+WqSX05y/+xsne0+L0tyTZLJJNuSPGXOa8we9ztJtrfrdyU5Y/Q/kfR4hoK0AEmOpXnS96iq+kWap6kf3N3+VXU/zcN1vzbwGr8APLN2PSF+QVW9mGaqjo+2T+BLvTIUpIVZDTxaVf8AUFWPVtX/3ssxl9I8oTprPc2UJj+lqu4FfkDzYKDUK0NBWpjPAmuTfD3JHyX5Fws45nLgtIF5mF5HExQ/pf1QmHurasfcbdKoLdv7LpKq6u+T/HOa4aCXA59sPyVrvikBqj3mkfZTuU5I8giws6oGP6XrnUnOpJlu+TeG+xNIC2MoSAtUVT8GbgJuSnI7zcR236YZ9nm03e2wgWXYNYT0CI8/S7igqj6Q5BTgoiTPr6r/N8QfQdorh4+kBUjywiRHDDS9mOYjIW8C3tDucxDwb2lmaZ11Jc2EZ69jnusJAFV1Nc2kgxvm2y6NkmcK0sI8A/jDNB9mv5Nm1s2NNDNkXpjkVprZOj9DM0smAFX1WJIvAj9XVfft4fX/K/DnSf64qn6yh/2koXKWVElSx+EjSVLHUJAkdQwFSVLHUJAkdQwFSVLHUJAkdQwFSVLn/wMFgoSwFaNe+gAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# loading information of covariates to numpy arrays.\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "base_pheno = pd.read_csv(\"phenotype_abetaratio.txt\",sep=\"\\t\")\n",
    "#abetaratio_np = base_pheno[\"abeta42\"].values\n",
    "abetaratio_np = base_pheno[\"abeta_ratio\"].values\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(1, 1, 1)\n",
    "ax.hist(abetaratio_np)\n",
    "ax.set_xlabel(\"Abeta 42/40 ratio\")\n",
    "ax.set_ylabel(\"Count\")\n",
    "plt.savefig(\"base_distribution\")\n",
    "\n",
    "base_cov = pd.read_csv(\"cov_agesexAPOE.txt\",sep=\"\\t\")\n",
    "base_cov_np = base_cov.loc[:,\"AGE\":\"APOEe4\"].values\n",
    "\n",
    "target_df_keep = []\n",
    "cnt=0\n",
    "for samp in target_phenosample_list:\n",
    "    if samp in target_VCFsample_list:\n",
    "        target_df_keep.append(cnt)\n",
    "    cnt = cnt+1\n",
    "if eval_phenotype==0:\n",
    "    target_pheno = pd.read_csv(\"pheno_merge_ADNI_SUVR.txt\",sep=\"\\t\")\n",
    "else:\n",
    "    target_pheno = pd.read_csv(\"pheno_merge_ADNI_CSF.txt\",sep=\"\\t\")\n",
    "target_pheno = target_pheno.loc[target_df_keep,:]\n",
    "\n",
    "if eval_phenotype==0:\n",
    "    target_pheno_np = target_pheno[\"AV45_bl\"].values\n",
    "else if eval_phenotype==1:\n",
    "    target_pheno_np = target_pheno[\"ABETA_bl\"].values\n",
    "else if eval_phenotype==2:\n",
    "    target_pheno_np = target_pheno[\"TAU_bl\"].values/target_pheno[\"ABETA_bl\"].values\n",
    "else if eval_phenotype==3: \n",
    "    target_pheno_np = target_pheno[\"PTAU_bl\"].values/target_pheno[\"ABETA_bl\"].values\n",
    "\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(1, 1, 1)\n",
    "ax.hist(target_pheno_np)\n",
    "ax.set_xlabel(\"SUVR\")\n",
    "ax.set_ylabel(\"Count\")\n",
    "plt.savefig(\"target_distribution\")\n",
    "\n",
    "if eval_phenotype==0:\n",
    "    target_cov = pd.read_csv(\"cov_merge_ADNI_SUVR.txt\",sep=\"\\t\")\n",
    "else:\n",
    "    target_cov = pd.read_csv(\"cov_merge_ADNI_CSF.txt\",sep=\"\\t\")\n",
    "target_cov = target_cov.loc[target_df_keep,:]\n",
    "target_cov_np = target_cov.loc[:,\"AGE\":\"APOE4\"].values\n",
    "\n",
    "print(target_cov_np.shape)\n",
    "print(target_pheno_np.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(102, 497)\n"
     ]
    }
   ],
   "source": [
    "# compensate missing genotype data with mean value\n",
    "\n",
    "if np.isnan(base_geno_np).any():\n",
    "    imp = SimpleImputer(missing_values=np.nan, strategy='mean', verbose=0, copy=True)\n",
    "    base_geno_np = imp.fit_transform(base_geno_np)  \n",
    "\n",
    "base_genocov_np = np.concatenate([base_geno_np,base_cov_np],axis=1)\n",
    "\n",
    "if np.isnan(target_geno_np).any():\n",
    "    imp = SimpleImputer(missing_values=np.nan, strategy='mean', verbose=0, copy=True)\n",
    "    target_geno_np = imp.fit_transform(target_geno_np)\n",
    "\n",
    "target_genocov_np = np.concatenate([target_geno_np,target_cov_np],axis=1)\n",
    "print(base_genocov_np.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "## fulfill missing genotypes in target data with mean of base's\n",
    "\n",
    "base_sampdim = base_geno_np.shape[0]\n",
    "base_vardim = base_geno_np.shape[1]\n",
    "target_sampdim = target_geno_np.shape[0]\n",
    "target_vardim = target_geno_np.shape[1]\n",
    "target_geno_basecoord = np.zeros([target_sampdim,base_vardim])\n",
    "targetvar_isinbase = np.zeros([base_vardim,1])\n",
    "base_cnt=0\n",
    "target_cnt=0\n",
    "for var in base_selected_var:\n",
    "    if var in target_selected_var:\n",
    "        targetvar_isinbase[base_cnt] = 1\n",
    "        target_geno_basecoord[:,base_cnt] = target_geno_np[:,target_cnt]\n",
    "        target_cnt = target_cnt+1\n",
    "    else:\n",
    "        #print(var)\n",
    "        targetvar_isinbase[base_cnt] = 0\n",
    "        target_geno_basecoord[:,base_cnt] = np.squeeze(np.mean(base_geno_np,axis=0)[base_cnt]*np.ones([target_sampdim,1]))\n",
    "        \n",
    "    base_cnt=base_cnt+1\n",
    "\n",
    "#print(base_cnt)\n",
    "#print(base_vardim)\n",
    "#print(target_cnt)\n",
    "#print(target_vardim)\n",
    "#print(target_geno_basecoord.shape)\n",
    "\n",
    "target_genocov_basecoord = np.concatenate([target_geno_basecoord,target_cov_np],axis=1)\n",
    "(target_genocov_basecoord.shape)\n",
    "targetvar_isinbase = np.squeeze(targetvar_isinbase)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:30: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  method='lar', copy_X=True, eps=np.finfo(np.float).eps,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:167: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  method='lar', copy_X=True, eps=np.finfo(np.float).eps,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:284: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  eps=np.finfo(np.float).eps, copy_Gram=True, verbose=0,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:862: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  eps=np.finfo(np.float).eps, copy_X=True, fit_path=True,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:1101: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  eps=np.finfo(np.float).eps, copy_X=True, fit_path=True,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:1127: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  eps=np.finfo(np.float).eps, positive=False):\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:1362: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  max_n_alphas=1000, n_jobs=None, eps=np.finfo(np.float).eps,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:1602: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  max_n_alphas=1000, n_jobs=None, eps=np.finfo(np.float).eps,\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/linear_model/least_angle.py:1738: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  eps=np.finfo(np.float).eps, copy_X=True, positive=False):\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/decomposition/online_lda.py:29: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  EPS = np.finfo(np.float).eps\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/ensemble/gradient_boosting.py:32: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  from ._gradient_boosting import predict_stages\n",
      "/Users/yuichiro/Workspace/Genome/ADabetaQTL/.env/lib/python3.7/site-packages/sklearn/ensemble/gradient_boosting.py:32: DeprecationWarning: `np.bool` is a deprecated alias for the builtin `bool`. To silence this warning, use `bool` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.bool_` here.\n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  from ._gradient_boosting import predict_stages\n"
     ]
    },
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'seaborn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-afb104e30640>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     12\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0msklearn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmetrics\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mroc_curve\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mroc_auc_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf1_score\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maccuracy_score\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mscipy\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mstats\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 14\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mseaborn\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     15\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpickle\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'seaborn'"
     ]
    }
   ],
   "source": [
    "#### Classification of Abeta postive/negative + iPSC eQTL & pca train -> ADNI pca test and model CV  %2020/5/21\n",
    "## + covariate only %2020/5/23\n",
    "\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "from sklearn.model_selection import GridSearchCV, cross_val_score, RepeatedKFold, KFold\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import roc_curve, roc_auc_score, f1_score, accuracy_score\n",
    "from scipy import stats\n",
    "import seaborn as sns\n",
    "import pickle\n",
    "\n",
    "if eval_phenotype==0:\n",
    "    Abeta_posneg_np = target_pheno_np>1.11\n",
    "else if eval_phenotype==1:\n",
    "    Abeta_posneg_np = (target_pheno_np<977.0)\n",
    "else if eval_phenotype==2:\n",
    "    Abeta_posneg_np = (target_pheno_np>0.27) #TAU\n",
    "else if eval_phenotype==3:\n",
    "    Abeta_posneg_np = (target_pheno_np>0.025) #pTAU\n",
    "acc_summary = []\n",
    "f1_summary = []\n",
    "auc_summary = []\n",
    "merge_tprs = []\n",
    "merge_aucs = []\n",
    "mean_merge_fpr = np.linspace(0, 1, 20)\n",
    "\n",
    "cov_acc_summary = []\n",
    "cov_f1_summary = []\n",
    "cov_auc_summary = []\n",
    "cov_merge_tprs = []\n",
    "cov_merge_aucs = []\n",
    "\n",
    "## PCA training\n",
    "base_x = base_geno_np\n",
    "n_components_pca = 3\n",
    "scaler = StandardScaler()\n",
    "base_x_scale = scaler.fit_transform(base_x)\n",
    "pca = PCA(svd_solver='full', n_components = n_components_pca)\n",
    "base_x_pca = pca.fit_transform(base_x_scale)\n",
    "plt.plot(pca.explained_variance_ratio_,'.')\n",
    "\n",
    "filename = 'ab_baseiPSCpc.pkl'\n",
    "pickle.dump(pca, open(filename, 'wb'))\n",
    "\n",
    "plt.figure(figsize=(10,100))\n",
    "df_pcacomponent = pd.DataFrame(data=np.transpose(pca.components_[:,targetvar_isinbase==1]),columns=['PC1','PC2','PC3'],index=target_selected_var)\n",
    "sns.heatmap(df_pcacomponent,cbar_kws={'label': 'Weight'})\n",
    "plt.savefig(\"pcacomponent.png\",format=\"png\",dpi=300)\n",
    "\n",
    "feat_importance_summary = np.zeros([10,6])\n",
    "\n",
    "cnt = 0\n",
    "kf = KFold(n_splits=10)\n",
    "for train_index, test_index in kf.split(target_geno_basecoord,Abeta_posneg_np):\n",
    "    y_train, y_test = Abeta_posneg_np[train_index], Abeta_posneg_np[test_index]\n",
    "    # genotype + covariate prediction model.\n",
    "    x_train, x_test = target_geno_basecoord[train_index], target_geno_basecoord[test_index]\n",
    "    x_train = scaler.transform(x_train)\n",
    "    x_test = scaler.transform(x_test)\n",
    "    x_train_pca = pca.transform(x_train)\n",
    "    x_test_pca = pca.transform(x_test)\n",
    "    x_train_pca = np.concatenate([x_train_pca,target_cov_np[train_index]],axis=1)\n",
    "    x_test_pca = np.concatenate([x_test_pca,target_cov_np[test_index]],axis=1)\n",
    "    classi = RandomForestClassifier(n_estimators=100)\n",
    "    classi.fit(x_train_pca, y_train)\n",
    "    y_pred = classi.predict(x_test_pca)\n",
    "    y_pred_prob = classi.predict_proba(x_test_pca)[:,1]\n",
    "    \n",
    "    fpr, tpr, thresholds = roc_curve(y_test, y_pred_prob, pos_label=1, drop_intermediate=False)\n",
    "    merge_tprs.append(np.interp(mean_merge_fpr, fpr, tpr))\n",
    "    merge_tprs[-1][0] = 0.0\n",
    "    auc_summary = np.append(auc_summary,roc_auc_score(y_test, y_pred_prob))\n",
    "    acc_summary = np.append(acc_summary,accuracy_score(y_test,y_pred))\n",
    "    f1_summary = np.append(f1_summary,f1_score(y_test,y_pred))\n",
    "    \n",
    "    feat_importance_summary[cnt] = classi.feature_importances_\n",
    "    \n",
    "    filename = 'ab_classifier_'+str(cnt)+'.pkl'\n",
    "    pickle.dump(classi, open(filename, 'wb'))\n",
    "    cnt = cnt + 1\n",
    "    \n",
    "    # only covariate prediction model.\n",
    "    x_train, x_test = target_cov_np[train_index], target_cov_np[test_index]\n",
    "    #cov_scaler = StandardScaler()\n",
    "    #x_train = cov_scaler.fit_transform(x_train)\n",
    "    #x_test = cov_scaler.transform(x_test)\n",
    "    cov_classi = RandomForestClassifier(n_estimators=100)\n",
    "    cov_classi.fit(x_train, y_train)\n",
    "    y_pred = cov_classi.predict(x_test)\n",
    "    y_pred_prob = cov_classi.predict_proba(x_test)[:,1]\n",
    "      \n",
    "    fpr, tpr, thresholds = roc_curve(y_test, y_pred_prob, pos_label=1, drop_intermediate=False)\n",
    "    cov_merge_tprs.append(np.interp(mean_merge_fpr, fpr, tpr))\n",
    "    cov_merge_tprs[-1][0] = 0.0\n",
    "    cov_auc_summary = np.append(cov_auc_summary,roc_auc_score(y_test, y_pred_prob))\n",
    "    cov_acc_summary = np.append(cov_acc_summary,accuracy_score(y_test,y_pred))\n",
    "    cov_f1_summary = np.append(cov_f1_summary,f1_score(y_test,y_pred))  \n",
    "    \n",
    "plt.figure(figsize=(9,4))\n",
    "df_featimportance = pd.DataFrame(data=np.transpose(feat_importance_summary),columns=['Set1','Set2','Set3','Set4','Set5','Set6','Set7','Set8','Set9','Set10'],index=['PC1','PC2','PC3','AGE','SEX','APOEe4'])\n",
    "sns.heatmap(df_featimportance,cbar_kws={'label': 'Importance'})\n",
    "plt.savefig(\"feat_imporance.png\",format=\"png\",dpi=300)\n",
    "\n",
    "## genotype + covariate\n",
    "# score for ROC\n",
    "merge_tprs = np.asarray(merge_tprs)\n",
    "mean_merge_tpr = np.mean(merge_tprs,axis=0)\n",
    "std_merge_tpr = np.std(merge_tprs,axis=0)\n",
    "\n",
    "fig = plt.figure(figsize=(4.0, 4.0))\n",
    "ax = fig.add_subplot(1, 1, 1)\n",
    "ax.plot([0,1],[0,1],'r--')\n",
    "ax.plot(mean_merge_fpr, mean_merge_tpr,'b-')\n",
    "ax.fill_between(mean_merge_fpr,mean_merge_tpr-std_merge_tpr,mean_merge_tpr+std_merge_tpr,alpha=0.3,linewidth=1,linestyle='dashed',edgecolor='k',facecolor='b')\n",
    "ax.set_xlabel('FPR')\n",
    "ax.set_ylabel('TPR')\n",
    "#plt.savefig(\"ROC_results.png\")\n",
    "plt.savefig('ROC_results', dpi=600)\n",
    "plt.show()\n",
    "print(np.mean(auc_summary))\n",
    "print(np.std(auc_summary))\n",
    "print(auc_summary)\n",
    "print(np.mean(acc_summary))\n",
    "print(np.std(acc_summary))\n",
    "print(acc_summary)\n",
    "print(np.mean(f1_summary))\n",
    "print(np.std(f1_summary))\n",
    "print(f1_summary)\n",
    "\n",
    "## only covariate\n",
    "# score for ROC\n",
    "cov_merge_tprs = np.asarray(cov_merge_tprs)\n",
    "cov_mean_merge_tpr = np.mean(cov_merge_tprs,axis=0)\n",
    "cov_std_merge_tpr = np.std(cov_merge_tprs,axis=0)\n",
    "\n",
    "fig = plt.figure(figsize=(4.0, 4.0))\n",
    "ax = fig.add_subplot(1, 1, 1)\n",
    "ax.plot([0,1],[0,1],'r--')\n",
    "ax.plot(mean_merge_fpr, cov_mean_merge_tpr,'b-')\n",
    "ax.fill_between(mean_merge_fpr,cov_mean_merge_tpr-cov_std_merge_tpr,cov_mean_merge_tpr+cov_std_merge_tpr,alpha=0.3,linewidth=1,linestyle='dashed',edgecolor='k',facecolor='b')\n",
    "ax.set_xlabel('FPR')\n",
    "ax.set_ylabel('TPR')\n",
    "#plt.savefig(\"ROC_results.png\")\n",
    "plt.savefig('cov_ROC_results', dpi=600)\n",
    "plt.show()\n",
    "print(np.mean(cov_auc_summary))\n",
    "print(np.std(cov_auc_summary))\n",
    "print(cov_auc_summary)\n",
    "print(np.mean(cov_acc_summary))\n",
    "print(np.std(cov_acc_summary))\n",
    "print(cov_acc_summary)\n",
    "print(np.mean(cov_f1_summary))\n",
    "print(np.std(cov_f1_summary))\n",
    "print(cov_f1_summary)\n",
    "\n",
    "# Wilcoxon signed-rank test\n",
    "print(stats.wilcoxon(auc_summary,cov_auc_summary))\n",
    "\n",
    "print(sum(Abeta_posneg_np))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=[5,3])\n",
    "plt.plot(pca.explained_variance_ratio_,'.')\n",
    "plt.xlabel(\"PC\")\n",
    "plt.xticks([0,1,2],labels=[\"1\",\"2\",\"3\"])\n",
    "plt.ylabel(\"Variance ratio\")\n",
    "plt.savefig(\"PCAexpvar.png\",format=\"png\",dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pca' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-19b3c18d4e2b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mmpl_toolkits\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmplot3d\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mAxes3D\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mtarget_pca\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpca\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtransform\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_geno_basecoord\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mfig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mplt\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfigure\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfigsize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m7\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m7\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0max\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_subplot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m111\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mprojection\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'3d'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscatter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtarget_pca\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mAbeta_posneg_np\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtarget_pca\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mAbeta_posneg_np\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mtarget_cov_np\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mAbeta_posneg_np\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pca' is not defined"
     ]
    }
   ],
   "source": [
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "target_pca = pca.transform(target_geno_basecoord)\n",
    "fig = plt.figure(figsize=[7,7])\n",
    "ax = fig.add_subplot(111,projection='3d')\n",
    "ax.scatter(target_pca[Abeta_posneg_np==1,0],target_pca[Abeta_posneg_np==1,1],target_cov_np[Abeta_posneg_np==1,0],s=5)\n",
    "ax.scatter(target_pca[Abeta_posneg_np==0,0],target_pca[Abeta_posneg_np==0,1],target_cov_np[Abeta_posneg_np==0,0],s=5)\n",
    "ax.view_init(200, 60)\n",
    "ax.set_xlabel(\"PC1\")\n",
    "ax.set_ylabel(\"PC2\")\n",
    "ax.set_zlabel(\"AGE\")\n",
    "plt.savefig(\"pca_targetval.png\",format=\"png\",dpi=300)\n",
    "print(target_pca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[69.6  0.   0. ]\n",
      " [70.9  0.   0. ]\n",
      " [65.   1.   0. ]\n",
      " ...\n",
      " [75.2  1.   0. ]\n",
      " [91.4  1.   0. ]\n",
      " [68.1  1.   0. ]]\n",
      "--------- Onset AGE ----------\n",
      "72.78123924268503\n",
      "7.223611302824563\n",
      "72.31730103806228\n",
      "7.350977030561295\n",
      "73.2404109589041\n",
      "7.065370679929781\n",
      "\n",
      "------------- SEX ---------------\n",
      "581\n",
      "141\n",
      "148\n",
      "123\n",
      "169\n",
      "\n",
      "------------ APOE4 ------------\n",
      "233\n",
      "54\n",
      "2\n",
      "116\n",
      "134\n",
      "42\n"
     ]
    }
   ],
   "source": [
    "############################\n",
    "# Summerize covariates \n",
    "# 2021.2.9\n",
    "\n",
    "if eval_phenotype==0:\n",
    "    Abeta_posneg_np = target_pheno_np>1.11\n",
    "else if eval_phenotype==1:\n",
    "    Abeta_posneg_np = (target_pheno_np<977.0)\n",
    "else if eval_phenotype==2:\n",
    "    Abeta_posneg_np = (target_pheno_np>0.27) #TAU\n",
    "else if eval_phenotype==3:\n",
    "    Abeta_posneg_np = (target_pheno_np>0.025) #pTAU\n",
    "    \n",
    "print(target_cov_np)\n",
    "\n",
    "print(\"--------- Onset AGE ----------\")\n",
    "print(np.mean(target_cov_np[:,0]))\n",
    "print(np.std(target_cov_np[:,0]))\n",
    "print(np.mean(target_cov_np[np.where(Abeta_posneg_np==0),0]))\n",
    "print(np.std(target_cov_np[np.where(Abeta_posneg_np==0),0]))\n",
    "print(np.mean(target_cov_np[np.where(Abeta_posneg_np==1),0]))\n",
    "print(np.std(target_cov_np[np.where(Abeta_posneg_np==1),0]))\n",
    "\n",
    "print(\"\")\n",
    "print(\"------------- SEX ---------------\") \n",
    "print(len(target_cov_np[:,1]))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==0),1]==0))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==0),1]==1))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==1),1]==0))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==1),1]==1))\n",
    "\n",
    "print(\"\")\n",
    "print(\"------------ APOE4 ------------\")\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==0),2]==0))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==0),2]==1))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==0),2]==2))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==1),2]==0))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==1),2]==1))\n",
    "print(np.count_nonzero(target_cov_np[np.where(Abeta_posneg_np==1),2]==2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target_cov_np.shape\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
